src/http/one/Tokenizer.h

```diff
@@ -9,68 +9,46 @@
 #ifndef SQUID_SRC_HTTP_ONE_TOKENIZER_H
 #define SQUID_SRC_HTTP_ONE_TOKENIZER_H
 
-#include "parser/Tokenizer.h"
+class SBuf;
+
+namespace Parser {
+class Tokenizer;
+}
 
 namespace Http {
 namespace One {
 
 /**
- * Lexical processor extended to tokenize HTTP/1.x syntax.
+ * Attempt to parse a (token / quoted-string ) lexical construct.
+ *
```

Comment:
Please document the return value meaning. For example:
```suggestion
 * \returns true (and sets the value) after successfully parsing the token prefix or the entire quoted string
 * \returns false (and leaves the value intact) if more data is needed to parse the token/string
 * \throws on syntax violations
```

Reply:
The added documentation is lying -- a `true` return value does not necessarily mean that we have parsed a (complete or whole) token. You have added a `!tok.atEnd()` check for that case, but that check does not work the way you think it does:

```
// Distinguish complete tokens from token prefixes:
// the presence of trailing non-token characters means
// a complete token.
!tok.atEnd();
```

You are correct that the presence of a trailing non-token character means that we have parsed a complete token. However, the reverse is not true -- the lack of a trailing non-token character does not necessarily mean that we have not parsed a complete token. If the caller has received a complete input sequence, then being `atEnd()` in this context means the complete token was successfully parsed.

While a complete input sequence may not end with a token-or-quoted-string value for the current tokenOrQuotedString() caller (because chunk extensions are followed by CRLF), you are not writing this function for that specific caller. It has to work correctly for all callers. Thus, this atEof() check belongs to specific callers that can use it.

Reply:
>  If the caller has received a complete input sequence, then being atEnd() in this context means the complete token was successfully parsed.

The complete input sequence may be 'complete' from the receiver of the data point of view (the sender closed the connection), but still be incomplete (e.g., the sender aborted while sending the data).

> you are not writing this function for that specific caller It has to work correctly for all callers.

I assume that the caller of this method should be specific anyway: it belongs to Http::One namespace and the callers, obviously, should belong either. Within this context (HTTP/1.1 messages), all tokens finally end up with whitespaces/line separators: if the input finished but still lacks them, the received token is invalid either.

Reply:
> The complete input sequence may be 'complete' from the receiver of the data point of view (the sender closed the connection), but still be incomplete (e.g., the sender aborted while sending the data).

Here, completeness is defined by this function caller. It does not matter to _this_ function what "complete" actually means -- it cannot (and does not want to) know how "completeness" is defined. You are right that the definition of completeness in the caller may be very complex, but that definition is caller-specific and irrelevant to this general tokenizing function.

Squid Tokenizer cannot tell whether "all data has been received" -- it simply lacks the API that can relay that information from the caller. Until that changes, Tokenizer and its knowledge-lacking users/wrappers (like this function) should not mistake an empty buffer for the promise of more token characters.

> callers, obviously, should belong either to the HTTP/1.1 messages context. Within this context, all tokens finally end up with whitespaces/line separators

No, we cannot assume that all callers are parsing an HTTP/1 message, and we cannot assume that each HTTP/1 token is followed by something else. HTTP/1 constructs such as tokens and quoted strings can be found embedded in other contexts where their length is determined by means other than whitespace and such. For a high-level example (at the whole message level), consider ICAP messages that embed HTTP headers and specify their exact size. And HTTP/2 reuses/embeds a lot of small HTTP/1 constructs, often specifying their exact size.

Reply:
Done.